### PR TITLE
Reorganize as a library with sub and iss support

### PR DESCRIPTION
**Pull request by lpabon**
Signed-off-by: Luis Pabón <luis@portworx.com>

### CODE REVIEW

**Comment by jrivera-px** [comment]
This is great, thanks for adding the auth pkg re-org 👍 !

File: pkg/auth/auth.go [hunk 1]
```diff
@@ -27,24 +27,29 @@ type Rule struct {
...
-	Key  interface{}
+	// (optional) RBAC rules for the OpenStorage SDK
+	Rules []Rule `json:"rules,omitempty" yaml:"rules,omitempty"`
 }
 
 // Options provide any options to apply to the token
 type Options struct {
+	// Expiration time in Unix format as per JWT standard
 	Expiration int64
+	// Set issuer in claims to this value
```
Comment (adityadani) [inline pkg/auth/auth.go]: nit: start the comment with Issuer or go vet is going to fail.

**Review by adityadani — APPROVED** [review]

File: pkg/auth/signature.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+// Signature describes the signature type using definitions from
+// the jwt package
+type Signature struct {
+	Type jwt.SigningMethod
+	Key  interface{}
+}
+
+func NewSignatureSharedSecret(secret string) (*Signature, error) {
```
Comment (ggriffiths) [inline pkg/auth/signature.go:32]: Maybe add some comments on these exported functions

File: pkg/auth/duration.go [hunk 3]
```diff
@@ -23,49 +23,49 @@ import (
...
-	yearRegex   = regexp.MustCompile("([0-9]+)" + yearDef)
+	SecondRegex = regexp.MustCompile("([0-9]+)" + SecondDef)
+	MinuteRegex = regexp.MustCompile("([0-9]+)" + MinuteDef)
+	HourRegex   = regexp.MustCompile("([0-9]+)" + HourDef)
+	DayRegex    = regexp.MustCompile("([0-9]+)" + DayDef)
+	YearRegex   = regexp.MustCompile("([0-9]+)" + YearDef)
 )
 
-func parseToDuration(s string) (time.Duration, error) {
+func ParseToDuration(s string) (time.Duration, error) {
```
Comment (ggriffiths) [inline pkg/auth/duration.go:47]: comment here as well

**Review by ggriffiths — APPROVED** [review]
LGTM, just some small comment suggestions

File: pkg/auth/auth.go [hunk 4]
```diff
@@ -54,7 +59,14 @@ func Token(
 	options *Options,
 ) (string, error) {
 
+	issuer := "openstorage-sdk-auth"
```
Comment (lpabon) [inline pkg/auth/auth.go]: issuer should be an option with a default.

**Pull request merged**